### PR TITLE
Cp for gradle 8.0 support in flutter 3.7

### PR DESCRIPTION
**Pull request by reidbaker**
Pull the minimal set of changes from flutter 3.10 (in beta) into flutter 3.7 to support gradle 8.0
[124838](https://github.com/flutter/flutter/issues/124838)

[Original PR](https://github.com/flutter/flutter/pull/121958)
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: dev/benchmarks/complex_layout/android/build.gradle [hunk 1]
```diff
@@ -14,7 +14,7 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:7.2.0'
+        classpath 'com.android.tools.build:gradle:7.3.0'
```
Comment (mossmana) [inline dev/benchmarks/complex_layout/android/build.gradle:17]: Out of curiosity, if there any reason why we couldn't start using a variable for version number like we do for the kotlin version?

Comment (reidbaker) [reply dev/benchmarks/complex_layout/android/build.gradle:17]: Agp version is more likely to cause you to have to make updates to your app than kotlin version or gradle version. Nothing mechanically stops us but I think leaving this version where it is easy for the Android Studio tooling to update is the best policy. That said it does probably make sense to have a version that can be shared for packages/packages.

File: dev/benchmarks/complex_layout/android/gradle/wrapper/gradle-wrapper.properties [hunk 2]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.0.2-bin.zip
```
Comment (gmackall) [inline dev/benchmarks/complex_layout/android/gradle/wrapper/gradle-wrapper.properties:3]: I notice this changes fron ***all.zip to ***bin.zip. Is there one we should generally prefer, or does it depend on the situation? 

As I understand, its the difference of whether sources/documentation come bundled, but I'm not sure what the implications of that for all these gradle wrappers in /benchmarks, for example.

Comment (reidbaker) [reply dev/benchmarks/complex_layout/android/gradle/wrapper/gradle-wrapper.properties:3]: I think for normal users all is probably safe since the size is not large. That said this is what is in the 3.10 branch and we dont want to make any changes.

File: dev/benchmarks/complex_layout/android/settings.gradle [hunk 3]
```diff
@@ -8,8 +8,6 @@
 
 include ':app'
 
-enableFeaturePreview('ONE_LOCKFILE_PER_PROJECT')
```
Comment (camsim99) [inline dev/benchmarks/complex_layout/android/settings.gradle:11]: Why was this removed? I

**Review by camsim99 — COMMENTED** [review]
The gradle changes look good, just had one question!

**Comment by reidbaker** [comment]
Mac builds are failing because `Exception on command "checkout e69253c6bd718486fdb5669a0c181900940e70f7"`. I think that is a an infra issue because my cherry pick is commit e69253c6bd718486fdb5669a0c181900940e70f7

File: dev/benchmarks/complex_layout/android/settings.gradle (hunk 3 above)
Comment (reidbaker) [reply dev/benchmarks/complex_layout/android/settings.gradle:11]: This was removed in the original pr because one lockfile per project was an experimental feature that was promoted to stable.

**Review by camsim99 — APPROVED** [review]
